### ISSUE TITLE
Hue dimmer level not updated (AC dimmers like Bush-Jaeger 6715u & Hue White Lamp )

### ISSUE THREAD

**Issue by marcovandeveen** [domoticz/domoticz#2613]
Changes in the dim level of a dimmer controlled by Philips Hue are not correctly reflected in Domoticz.  

It's not a normal Philips dimmer, but a Busch Jaeger 6715 u physical zigbee dimmer. Which is identified in Domoticz as
- Type: Lightning 2
- Subtype: AC

What does work:
- Dim level changes from Domoticz to the Busch & Jaeger dimmer in Philips Hue 
- On/off changes of the dimmer in Philips hue are correctly reflected in Domoticz. 

Version: 4.9796

![image](https://user-images.githubusercontent.com/42120041/43688486-0caebdec-98ea-11e8-99fb-3cb05ad3227e.png)
![buschjaeger6715udimmerindomoticz_2](https://user-images.githubusercontent.com/42120041/43688461-a1a5264e-98e9-11e8-883e-c0644694150a.jpg)

**Comment by chrislader** [comment]
I think its the same problem as in #2519.

The percentage of my hue lux lights (also identified as AC; which is correct since they can only dim in the same color;2700 K)

It seems that the polling of AC hue dimmers is not correctly done by domoticz. The other way around works; so on/off and percentage by domoticz (displayd correctly in Philips hue app).

**Comment by chrislader** [comment]
Sorry its #2529

**Comment by marcovandeveen** [comment]
Functions I use to keep dim levels in sync in the time being: 

    getHueLights = function(domoticz)
           local hueBridgeIP = 'xxxx'
           local hueBridgeAPI = xxxxi'
           local http = require('socket.http')
           local ltn12 = require('ltn12')
           local json = require('dkjson')
         
           t = {}
           local url = string.format("http://%s/api/%s/lights", hueBridgeIP, hueBridgeAPI)
           b, c, h = http.request{url=url, sink = ltn12.sink.table(t), method='GET'}
           huestring = tostring(table.concat(t))
           local hue, pos, err = json.decode(huestring, 1, nil)
           --domoticz.log('Successfully retrieved hue lights', domoticz.LOG_INFO)
           return hue
        end,
        syncBuschJaegerDimLevels = function(domoticz)
            local hueLights = domoticz.helpers.getHueLights(domoticz)
            for i, hueLight in pairs(hueLights) do
                if(hueLight.manufacturername =='Busch-Jaeger') then
                    local hueDimLevel = math.floor(hueLight.state.bri / 254 * 100)
                    local domoticzLight = domoticz.devices(hueLight.name)
                    if (domoticzLight.state == 'On' and hueDimLevel ~= domoticzLight.level)  then
                        domoticzLight.dimTo(hueDimLevel)
                        domoticz.log('Busch & Jaeger light ' .. hueLight.name .. '  synced from hue with domoticz', domoticz.LOG_INFO)
                    end
                end
            end
            domoticz.log('Synced all Busch & Jaeger lights with domoticz', domoticz.LOG_INFO)
        end

**Comment by gizmocuz** [comment]
@marcovandeveen , could you check the c++ code to see if something goes wrong there ?

**Comment by emontnemery** [comment]
@marcovandeveen, @chrislader: What OS are you running Domoticz on?

**Comment by chrislader** [comment]
Im running linux, stretch on a rasp pi 3B

**Comment by marcovandeveen** [comment]
Raspbian GNU/Linux 8 (jessie) 
Raspberry Pi 3 B

**Comment by marcovandeveen** [comment]
```
//Hue API response for Busch & Jaeger Dimmer which dim level is NOT updated in domoticz
{
	"state": {
		"on": false,
		"bri": 119,
		"alert": "none",
		"mode": "homeautomation",
		"reachable": true
	},
	"swupdate": {
		"state": "notupdatable",
		"lastinstall": null
	},
	"type": "Dimmable light",
	"name": "Kamer verlichting",
	"modelid": "RM01",
	"manufacturername": "Busch-Jaeger",
	"productname": "Dimmable light",
	"capabilities": {
		"certified": false,
		"control": {},
		"streaming": {
			"renderer": false,
			"proxy": false
		}
	},
	"config": {
		"archetype": "classicbulb",
		"function": "functional",
		"direction": "omnidirectional"
	},
	"uniqueid": "d8:5d:ef:11:a1:00:36:a0-12",
	"swversion": "1.2.0"
}

//Hue API response for native Hue lamp which dim level is NOT updated in domoticz
{
        "state": {
            "on": false,
            "bri": 102,
            "alert": "none",
            "mode": "homeautomation",
            "reachable": true
        },
        "swupdate": {
            "state": "noupdates",
            "lastinstall": "2017-11-19T13:10:56"
        },
        "type": "Dimmable light",
        "name": "Buitenlamp deeldeur",
        "modelid": "LWB010",
        "manufacturername": "Philips",
        "productname": "Hue white lamp",
        "capabilities": {
            "certified": true,
            "control": {
                "mindimlevel": 5000,
                "maxlumen": 806
            },
            "streaming": {
                "renderer": false,
                "proxy": false
            }
        },
        "config": {
            "archetype": "classicbulb",
            "function": "functional",
            "direction": "omnidirectional"
        },
        "uniqueid": "00:17:88:01:02:56:3a:05-0b",
        "swversion": "1.29.0_r21169",
        "swconfigid": "C40F3F07",
        "productid": "Philips-LWB010-1-A19DLv3"
    }

// //Hue API response for native Hue lamp which dim level is updated correctly in domoticz
{
        "state": {
            "on": true,
            "bri": 80,
            "hue": 63392,
            "sat": 254,
            "effect": "none",
            "xy": [
                0.6289,
                0.278
            ],
            "ct": 153,
            "alert": "none",
            "colormode": "xy",
            "mode": "homeautomation",
            "reachable": true
        },
        "swupdate": {
            "state": "noupdates",
            "lastinstall": "2017-11-15T13:45:22"
        },
        "type": "Extended color light",
        "name": "Mirthe",
        "modelid": "LCT012",
        "manufacturername": "Philips",
        "productname": "Hue color candle",
        "capabilities": {
            "certified": true,
            "control": {
                "mindimlevel": 2000,
                "maxlumen": 450,
                "colorgamuttype": "C",
                "colorgamut": [
                    [
                        0.6915,
                        0.3083
                    ],
                    [
                        0.17,
                        0.7
                    ],
                    [
                        0.1532,
                        0.0475
                    ]
                ],
                "ct": {
                    "min": 153,
                    "max": 500
                }
            },
            "streaming": {
                "renderer": true,
                "proxy": true
            }
        },
        "config": {
            "archetype": "candlebulb",
            "function": "mixed",
            "direction": "omnidirectional"
        },
        "uniqueid": "00:17:88:01:02:1c:31:05-0b",
        "swversion": "1.29.0_r21169",
        "swconfigid": "EA9279F9",
        "productid": "Philips-LCT012-1-E14ECLv1"
    },
```

**Comment by emontnemery** [comment]
@marcovandeveen Thanks a lot, that's extremely helpful! Do you build domoticz yourself, if not I'll prepare a debug/test build.

**Comment by marcovandeveen** [comment]
It's not related to Busch & Jaeger. Domoticz doesn't update the dim level correctly for some Hue native lights as well.

However other hue native lights do work correctly. Comparing the JSON it's probably related with the type:
- correctly updated:  `"type": "Extended color light"`
- not correctly updated: `"type": "Dimmable light"`

**Comment by marcovandeveen** [comment]
@emontnemery I'm also looking at C++ code, but I'm not programmer. I can do some scripting in LUA, but that's about it.  Also I have never build domoticz.

Maybe I should learn some c++  :-)

**Comment by chrislader** [comment]
I have 3 type of hue lamps;

- hue color, gen1, identified as type “RGBWW”. They work correctly
- hue white ambiance, identified as type “WW”. They work as well
- hue lux (older type), identified as type “AC”. They do work with on/off status. But when dimlevel is changed by using philips hue app, the level is not matched/synced in Domoticz

**Comment by chrislader** [comment]
![Image](https://i.imgur.com/J1hYVsq.jpg)

**Comment by emontnemery** [comment]
Please try this build: http://montnemery.com/domoticz_hue_dimlevel/001/domoticz_linux_armv7l.tgz

Dimlevel just might work, and if it doesn't, there are at least some additional debug prints :)

Please make a backup of your domoticz installation / database before testing.
Easiest way to test is to simply drop the domoticz binary from the domoticz_linux_armv7l.tgz in your domoticz install directory after making a backup of the domoticz database.

**Comment by marcovandeveen** [comment]
Ok. Will do a.s.a.p. and let you know the results.

**Comment by marcovandeveen** [comment]
Problem still persists

Build number: V4.9863 / Build Hash: 716a5255-modified.

Log file states: 
```
2018-08-16 21:13:30.810 Status: HueBridge Processing light with key: '1'
2018-08-16 21:13:30.810 Status: HueBridge Cached state on: 1, level: 100
2018-08-16 21:13:30.810 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 100
2018-08-16 21:13:30.810 Status: HueBridge Processing light with key: '11'
2018-08-16 21:13:30.810 Status: HueBridge Cached state on: 1, level: 31
2018-08-16 21:13:30.810 Status: HueBridge bDoSend: 1869605768, LType: 5, on: 1, level: 31
2018-08-16 21:13:30.810 Status: HueBridge Processing light with key: '12'
2018-08-16 21:13:30.810 Status: HueBridge Cached state on: 1, level: 64
2018-08-16 21:13:30.810 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 64
2018-08-16 21:13:30.810 Status: HueBridge Processing light with key: '13'
2018-08-16 21:13:30.810 Status: HueBridge Cached state on: 1, level: 66
2018-08-16 21:13:30.810 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 66
2018-08-16 21:13:30.810 Status: HueBridge Processing light with key: '2'
2018-08-16 21:13:30.810 Status: HueBridge Cached state on: 1, level: 76
2018-08-16 21:13:30.810 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 76
2018-08-16 21:13:30.811 Status: HueBridge Processing light with key: '3'
2018-08-16 21:13:30.811 Status: HueBridge Cached state on: 0, level: 49
2018-08-16 21:13:30.811 Status: HueBridge bDoSend: 1869605768, LType: 4, on: 0, level: 49
2018-08-16 21:13:30.811 Status: HueBridge Processing light with key: '4'
2018-08-16 21:13:30.811 Status: HueBridge Cached state on: 1, level: 31
2018-08-16 21:13:30.811 Status: HueBridge bDoSend: 1869605768, LType: 5, on: 1, level: 31
2018-08-16 21:13:30.811 Status: HueBridge Processing light with key: '5'
2018-08-16 21:13:30.811 Status: HueBridge Cached state on: 1, level: 100
2018-08-16 21:13:30.811 Status: HueBridge bDoSend: 1869605768, LType: 4, on: 1, level: 100
2018-08-16 21:13:30.811 Status: HueBridge Processing light with key: '8'
2018-08-16 21:13:30.811 Status: HueBridge Cached state on: 1, level: 0
2018-08-16 21:13:30.811 Status: HueBridge bDoSend: 1869605768, LType: 0, on: 1, level: 0
2018-08-16 21:13:30.811 Status: HueBridge Processing light with key: '9'
2018-08-16 21:13:30.811 Status: HueBridge Cached state on: 0, level: 41
2018-08-16 21:13:30.811 Status: HueBridge bDoSend: 1869605768, LType: 5, on: 0, level: 41
```

Related Hue API Response: https://github.com/domoticz/domoticz/files/2294949/HueAPIResponse.txt

**Comment by marcovandeveen** [comment]
TEST 1
I change light 2 (Philips Hue white lamp, model LWB010) in the HUE API to 89%
I see in the log file
 ```
2018-16 21:31:56.019 Status: HueBridge Processing light with key: '2'
2018-16 21:31:56.019 Status: HueBridge Cached state on: 1, level: 89
2018-16 21:31:56.019 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 89
```
I DO NOT see a change in the dim level in domoticz

TEST 2
I change light 2 (Philips Hue white lamp, model LWB010) in the HUE API to 19%
I see in the log file
 ```
2018-08-16 21:35:48.953 Status: HueBridge Processing light with key: '2'
2018-08-16 21:35:48.953 Status: HueBridge Cached state on: 1, level: 19
2018-08-16 21:35:48.953 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 19
 ```
I DO NOT see a change in the dim level in domoticz

TEST 3
I change light 4 (Philips Hue color candle, model LCT012) in the HUE API to 41%
I see in the log file
 ```
2018-08-16 21:42:31.059 Status: HueBridge Processing light with key: '4'
2018-08-16 21:42:31.059 Status: HueBridge Cached state on: 1, level: 41
2018-08-16 21:42:31.059 Status: HueBridge bDoSend: 1869605768, LType: 5, on: 1, level: 41
 ```
I DO see a change in the dim level in domoticz

TEST 4
I change light 13 (Busch-Jaeger  Dimmable light, model RM01) in the HUE API to 1%
I see in the log file
 ```
2018-08-16 21:47:19.789 Status: HueBridge Processing light with key: '13'
2018-08-16 21:47:19.789 Status: HueBridge Cached state on: 1, level: 1
2018-08-16 21:47:19.789 Status: HueBridge bDoSend: 1869605768, LType: 1, on: 1, level: 1
 ```
I DO NOT see a change in the dim level in domoticz

**Comment by chrislader** [comment]
LTW010 works (hue white ambiance)
LTW001 works (hue white ambiance)
LCT007 works (hue color gen1)

LWB004 doesnt (hue lux)

**Comment by emontnemery** [comment]
@marcovandeveen Thanks a lot for the log!
There were multiple bugs in the PhilipsHue code preventing simple "AC" dimmer from updating dimlevel.

Please try this build http://montnemery.com/domoticz_hue_dimlevel/002/domoticz_linux_armv7l.tgz

If dimlevel is now updating, please test if Domotic state is correct when:
- Turn on
- Turn off
- Changing dim level
- Change dim level to minimum
- Change dim level to maximum

**Comment by marcovandeveen** [comment]
Dim level is updated automatically in Domoticz!!

I did disable my custom lua script to update the level

I do see an issue in the dim percentage: 
1% in hue appl = bri 1 in hue api = 6% in domoticz 
15% in hue appl = bri 38 in hue api = 13% in domoticz
30% in hue appl = bri 76 in hue api = 26% in domoticz
50% in hue appl = bri 127 in hue api = 53 % in domoticz
98% in hue appl = bri 248 in hue api = 100% in domoticz
100% in hue appl = bri 254 in hue api= 100% in domoticz

I'll do full test with all lights later.

**Comment by emontnemery** [comment]
The issue with dim percentage is because the Philips Hue integration is using LIGHTING2 type to represent simple dimmers, which is limited to 16 dim levels.
This can be fixed, but I would suggest to handle that separately (maybe open an issue saying Philips Hue dimmer is limited to 16 steps).

**Comment by emontnemery** [comment]
@marcovandeveen Will you do more testing, or do you think the changes are good and can be merged?

**Comment by marcovandeveen** [comment]
I logged another issue for the dim level limitation: #2659

Changes seem good:
- I did the test you requested on the lights that were failing in my house: the bug has been resolved. 
- I don't see negative consequences from the the fix: The hue lights that were not failing, still work correctly as before. 

Thanks!

I'll do another round of testing in the beta after the merge. 

@chrislader: thanks for helping in validation
